Add Support for BMS Tiltback

Relys opened the pull request:
```
    (var bms-fault-codes '(
        (NONE . 0)
        (BMS_COMM_TIMEOUT . 1) < 2 SECONDS. Not currently used in Refloat. Not sure where to put to notify user.
        (BMS_OVER_TEMP . 2) CELL_OVER_TEMP +10
        (CELL_OVER_VOLTAGE . 3)  bms-vmax-limit-start
        (CELL_UNDER_VOLTAGE . 4) bms-vmin-limit-start
        (CELL_OVER_TEMP . 5) bms-t-limit-start
        (CELL_UNDER_TEMP . 6) ; Not used
        (CELL_BALANCE . 7) <= 0.5v difference. Not currently used in Refloat. Not sure how to alert user.
    ))

The other values are from conf-get function in lisp. They can be found in Motor Config -> BMS. While we want the limit boxes unchecked, we can use these fields to customize limits for our own use. Note: I also check if any limits are enabled and disable them (since we want to do that with pushback not limiting motor current).
'bms-t-limit-start      ; Temperature where current starts to get reduced
'bms-vmin-limit-start   ; VCell where current starts to get reduced
'bms-vmax-limit-start   ; VCell where regen current starts to get reduced
```

Now package.lisp spawns a new process bms-state-loop which handles looking at user config values and BMS data to determine BMS state. It will search for a bms-can-id within first 10 seconds of boot and load the bms-state library if found. BMS faults are stored in an u32 as bitfields. They are sent to the main Refloat C loop with ext-bms-set-fault. In the areas that handle pushback bitfield checks with bool bms_is_fault_set(uint32_t fault_mask, BMS_FAULT_CODES fault_code) are now performed. Minimal code was touched in Refloat itself besides the additional checks, and BeepReason was expanded to handle additional error states.

lukash commented on package.lisp:
```diff
@@ -7,6 +7,14 @@
 
 ; Set firmware version:
 (apply ext-set-fw-version (sysinfo 'fw-ver))
+(if (>= (+ (first (sysinfo 'fw-ver)) (* (second (sysinfo 'fw-ver)) 0.01)) 6.05) {
```
Could you please do something like this instead?
```lisp
(setq version_major (first (sysinfo 'fw-ver)))
(setq version_minor (second (sysinfo 'fw-ver)))
(if (or (>= version_major 7) (and (>= version_major 6) (>= version_minor 5))) {
```

Relys replied:
Waist of flash space IMO, but I changed it anyways.

lukash commented on package.lisp:
```diff
@@ -7,6 +7,14 @@
 ; Set firmware version:
 (apply ext-set-fw-version (sysinfo 'fw-ver))
+(if (>= (+ (first (sysinfo 'fw-ver)) (* (second (sysinfo 'fw-ver)) 0.01)) 6.05) {
+    (loopwhile (or (< (get-bms-val 'bms-can-id) 0) (< (secs-since 0) 10)) (sleep 1))
```
What's the unit for `(sleep 1)`? 1 second?

Relys replied:
Yes it's 1 second. This is a standard function documented in Lisp documentation manuals.

lukash commented on package.lisp:
```diff
@@ -7,6 +7,14 @@
 (apply ext-set-fw-version (sysinfo 'fw-ver))
+(if (>= (+ (first (sysinfo 'fw-ver)) (* (second (sysinfo 'fw-ver)) 0.01)) 6.05) {
+    (loopwhile (or (< (get-bms-val 'bms-can-id) 0) (< (secs-since 0) 10)) (sleep 1))
+    (if (>= (get-bms-val 'bms-can-id) 0){
```
Correct me if I'm wrong, but isn't the `{` syntax 6.05+? Does this work on 6.02?

Also, coding-style nitpick, space between `)` and `{`. Please go through the PR and unify these if you can.

Relys replied:
No, I've been using { syntax in 6.02. It's just a cleaner alternative to progn.

Space between ) { is what's used in all lisp examples and is standard coding-style for lisp.

lukash replied:
Well, you're missing the space on this line, that's why I'm mentioning it.

Relys replied:
ah i see now.

lukash commented on src/bms-state.lisp:
```diff
@@ -0,0 +1,70 @@
+(defun bms-state-loop () {
+    (var bms-vmin-limit-start (conf-get 'bms-vmin-limit-start))
+    (var bms-vmax-limit-start (conf-get 'bms-vmax-limit-start))
+    (var bms-tmax-limit-start (- (conf-get 'bms-t-limit-start) 3)) ; start 3 degrees before we start limiting
```
Nitpick, but the comment doesn't make much sense, either phrase it better or just remove it please.

Relys replied:
Rephrased. We set it lower than what's the value in the Motor CFG -> BMS so we get warning before we start limiting current if that option is selected. @surfdado recommended this.

lukash commented on src/bms-state.lisp:
```diff
@@ -0,0 +1,70 @@
+    (var bms-vmax-limit-start (conf-get 'bms-vmax-limit-start))
+    (var bms-tmax-limit-start (- (conf-get 'bms-t-limit-start) 3)) ; start 3 degrees before we start limiting
+    (var bms-cell-balance-start 0.5)
+    (var bms-discharge-tmax-limit-start (+ bms-tmax-limit-start 10)) ; set bms mosfet discharge limit to +10 deg c past cells
```
This comment I completely fail to understand though :smiley:

Relys replied:
I can rephrase this. I guess I was thinking for a BMS used that has discharge path, so I was gonna check the BMS temp sensor too.

Relys replied:
Going to increase to + 15 over cell value for BMS IC.

lukash commented on src/bms-state.lisp:
```diff
@@ -0,0 +1,70 @@
+    (var bms-config-update-time (systime))
+    (var bms-fault 0u32)
+
+    (var v-cell-support (>= (+ (first (sysinfo 'fw-ver)) (* (second (sysinfo 'fw-ver)) 0.01)) 6.06))
```
Ditto here with the version check, please compare numbers directly.

lukash commented on src/bms-state.lisp:
```diff
@@ -0,0 +1,70 @@
+    (var bms-tmax-limit-start (- (conf-get 'bms-t-limit-start) 3)) ; start 3 degrees before we start limiting
+    (var bms-cell-balance-start 0.5)
+    (var bms-discharge-tmax-limit-start (+ bms-tmax-limit-start 10)) ; set bms mosfet discharge limit to +10 deg c past cells
+    (var bms-config-update-time (systime))
```
It looks these are duplicated inside the main loop below, can they be removed? If these lines are for some reason necessary, please put them in a function and call it from the two places.

lukash commented on src/bms-state.lisp:
```diff
@@ -0,0 +1,70 @@
+    (var bms-vmin-limit-start (conf-get 'bms-vmin-limit-start))
+    (var bms-vmax-limit-start (conf-get 'bms-vmax-limit-start))
+    (var bms-tmax-limit-start (- (conf-get 'bms-t-limit-start) 3)) ; start 3 degrees before we start limiting
+    (var bms-cell-balance-start 0.5)
```
0.5V is too much for sure. You had 0.05V in the previous version. Did we get this number confirmed by some battery people?

lukash commented on src/bms-state.lisp:
```diff
@@ -0,0 +1,70 @@
+            (if (>= (get-bms-val 'bms-temp-ic) bms-discharge-tmax-limit-start) (bms-set-fault 'BMS_OVER_TEMP))
+        })
+        (apply ext-bms-set-fault bms-fault)
+        (yield 50000)
```
Is this 50ms? Can you make it 100? At what rate do the bms values update? (I assume they are being read from memory, which is being updated as the messages come from the bms)

lukash commented on src/bms-state.lisp:
```diff
@@ -0,0 +1,70 @@
+        (apply ext-bms-set-fault bms-fault)
+        (yield 50000)
+    })
+})
```
Trailing newline missing, please add it.

Relys replied:
Added.

lukash commented on src/bms-state.lisp:
```diff
@@ -0,0 +1,70 @@
+    })
+
+    (loopwhile t {
+        (if (and (<= (get-duty) 0.05)  (> (secs-since bms-config-update-time) 1.0)) { ; Only bother updating config values while board is idle
```
I'm not too keen on the duty > 0.05 check, looks somewhat arbitrary. Stuff like this tends to strike back at some point. e.g., what if you set your board to flywheel and decide to finally fix your bms thresholds? And somehow the flywheel never goes under 0.05 duty (I have no clue if it actually does or not).

Wouldn't there per chance be some sort of a callback on the API to notify of config changes?

Relys replied:
d->motor.duty_cycle > 0.05 is used in main.c for determining if board is should do HV/LV tiltback.

If there's a callback on the API for config changes let me know.

If someone was in flywheel mode and changed a config to fix bms thresholds (why you'd want to even change these from defaults I have no clue), the value would be updated as soon as 1 second after it's disengaged.

lukash replied:
> d->motor.duty_cycle > 0.05 is used in main.c for determining if board is should do HV/LV tiltback.

Yes. That's quite a different purpose for which it's used, I don't know how you draw the connection to here and use the same mechanism.

> If there's a callback on the API for config changes let me know.

Ah. I thought you'd be more familiar with the interface and it'd be easier for you to find.

I had a look and didn't find it.

> If someone was in flywheel mode and changed a config to fix bms thresholds (why you'd want to even change these from defaults I have no clue), the value would be updated as soon as 1 second after it's disengaged.

You didn't understand the example situation. I was trying to point out the values could have been needed before disengaging the flywheel.

I think you could just poll for the config changes every second and it'll be fine, though not optimal...

Relys replied:
poll every second

lukash commented on src/conf/datatypes.h:
```diff
@@ -21,6 +21,17 @@
+    CELL_OVER_TEMP = 5,
+    CELL_UNDER_TEMP = 6,
+    CELL_BALANCE = 7
+} BMS_FAULT_CODES;
```
This enum doesn't belong here, this header is for conf data types only. I understand this is very little code, but I think just add a bms module (bms.h and bms.c) anyway.

Also, please use `CamelCase` for data type names and add a common prefix to all the enum items, as they are in the global namespace.

Relys replied:
Put in bms.h and bms.c and used camel case.

lukash commented on src/utils.c:
```diff
@@ -37,3 +37,10 @@ float clampf(float value, float min, float max) {
+        return false;
+    }
+    return (fault_mask & (1U << (fault_code - 1))) != 0;
+}
```
This function also doesn't belong to the utils module (see comment about `BMS_FAULT_CODES`).

Relys replied:
put in bms.h bms.c

lukash commented on ui.qml.in:
```diff
@@ -867,6 +867,13 @@ Item {
+            [14, "Cell High Voltage"],
+            [15, "Cell Balance"],
+            [16, "BMS Connection"],
+            [17, "BMS Temp"],
```
I wonder if all these should be prefixed with "BMS: " just to make it clear where it comes from. For the last two items specifically:
> BMS: Connection Lost
BMS: BMS Temp

 (yeah, the double BMS is not very pretty on the last one but I don't have a better idea)

Relys replied:
The beep reason is already quite long as far as display in UI and I don't think there's ambiguity as to what Cell means.

lukash replied:
I disagree, I think for casual users clearly stating it's an error coming from BMS is important and helpful. There's room in the UI. I'd like all the warnings coming from the BMS to be clearly differentiated.

Relys replied:
ok

lukash reviewed (changes requested):
You're not handling the Cell Balance and Comm Timeout faults on the C side and you removed the Cell Under Temp handling altogether (I feel this should be addressed in some way).

The commit message is chaotic. Is it really signed off by @surfdado? Please add a `Feature: ` git trailer for the changelog.

Relys commented:
Added some changes, but am unable to test right now since my board is down.

I added beeps (no tiltback) for balance and communication errors. I also re-added CELL_UNDER_TEMP when cell goes below 0 degrees (as per recommended by Battery Mooch).

Instead of keeping logic where I loop through bms voltages, I was able to get accessing the min/max cell values through lisp pulled upstream into main bldc branch: https://github.com/vedderb/bldc/commit/07128eaebbec7433a7dfb37c8a44fab0090425c7

Instead of hard coding a specific version, I can just check if that function exists before continuing: 

`(if (eq (first (trap (get-bms-val 'bms-v-cell-min))) 'exit-ok)`

lukash commented on src/bms.h:
```diff
@@ -0,0 +1,34 @@
+#include <stdint.h>
+
+typedef enum {
+    NONE = 0,
```
`NONE` needs the `BMS_` prefix the most, as it has the highest chance to conflict with another symbol.

What about making the prefix `BMSF_` for a bit more disambiguation?

Relys replied:
done

lukash commented on src/bms.lisp:
```diff
@@ -0,0 +1,49 @@
+    (var bms-vmax-limit-start (conf-get 'bms-vmax-limit-start))
+    (var bms-tmax-limit-start (- (conf-get 'bms-t-limit-start) 3)) ; Start 3 degrees before Motor CFG -> BMS limiting functionality would happen.
+    (var bms-tmin-limit-start 0)
+    (var bms-cell-balance-start 0.5)
```
Unaddressed comment from last review, 0.5 seems way too high to me. Not sure what the value should be though.

Relys replied:
That's what dado recommended. Cells dip especially on split packet and where there's noise.

I'm wondering if I should just remove this entirely? Or maybe only beep when the board is not running and then we could lower that threshold?

lukash replied:
Oh, I see. True that properly judging how the pack is balanced should only be done when idle, it can jump around a lot while riding. We could check it only when idle, question is what to do, is the board supposed to beep in that case? Is it supposed to beep continuously or once in a while? Maybe seeing the battery is imbalanced in the UI would be enough? But then, the dedicated BMS page in VESC Tool might be a better place for that...

FTR, propagating e.g. the `engaged` state to lisp and moving more of the logic over to there is exactly what I _don't_ want to do.

So this brings me around to whether we should do anything at all, and if we should do anything, seems just checking the balance is not totally out of whack, like, having a 0.5V difference, makes some sense to me :grinning: .

In that case, please add a comment along the lines of:
```
// Only check if the balance is not completely off (in idle or while riding), not meant as a proper balance check
```

Relys replied:
Right now I'm only checking if the board state is READY (not RUNNING) in main.c: https://github.com/lukash/refloat/pull/27/files#diff-e0cf5b28d9b6b600f0af2bc78e8fd30ec675fd731a5da86f0c4283ffc0e40176R1497

Same for BMS connection issue. Should BMS connection actually be checked while running though, and balance only alert while idle?

lukash commented on src/bms.lisp:
```diff
@@ -0,0 +1,49 @@
+    (var bms-tmax-limit-start (- (conf-get 'bms-t-limit-start) 3)) ; Start 3 degrees before Motor CFG -> BMS limiting functionality would happen.
+    (var bms-tmin-limit-start 0)
+    (var bms-cell-balance-start 0.5)
+    (var bms-ic-tmax-limit-start (+ bms-tmax-limit-start 15)) ; Set bms temp limit to +15C past cell limit
```
Just a note, looking at this 15 degree constant being pretty arbitrary. We should at least figure out a place to document these numbers so that users know how it works...

(I don't like the arbitrary number but don't have a better suggestion, seems reasonable at least to start with)

Relys replied:
yeah idk +15 just seems like a reasonable overhead. for cell max 45 +15 would be 60 on bms (and bms is next to cell pack so you don't want it getting too hot) I could also just not look at this value.

lukash replied:
I don't know, makes some sense to warn about this, if someone's discharge bms starts catching fire... I think we just need to set up some documentation page for this stuff...

lukash commented on src/bms.lisp:
```diff
@@ -0,0 +1,49 @@
+    (var bms-tmax-limit-start (- (conf-get 'bms-t-limit-start) 3)) ; Start 3 degrees before Motor CFG -> BMS limiting functionality would happen.
+    (var bms-tmin-limit-start 0)
+    (var bms-cell-balance-start 0.5)
+    (var bms-ic-tmax-limit-start (+ bms-tmax-limit-start 15)) ; Set bms temp limit to +15C past cell limit
```
Could you name the variables (on lines 2-7) in a better way? the `bms-` prefix is redundant in this context. The `-start` sufix as well. What about something like:
```
v-cell-max-limit
v-cell-min-limit
temp-cell-max-limit
...
```

Relys replied:
yeah that works

lukash commented on src/bms.lisp:
```diff
@@ -0,0 +1,49 @@
+            (setq bms-tmax-limit-start (- (conf-get 'bms-t-limit-start) 3))
+            (setq bms-ic-tmax-limit-start (+ bms-tmax-limit-start 15))
+            (setq bms-config-update-time (systime))
+        })
```
Unaddressed from previous review, this code is duplicate with the one on lines 2-7. Please move it out to a function.

Relys replied:
Technically this is not duplicate. Lines 2-7 are variable declaration (var) and initial assignment. I'll move it to a separately function, but technically that's just creating more overhead. Lisp is not a compiled language...

lukash replied:
An overhead of a single function call? Unless that's more expensive than I though, I just consider it a necessity for well-written code, it's significantly better from maintenance perspective to not duplicate e.g. the 3 and the 15 constants.

lukash commented on src/bms.lisp:
```diff
@@ -0,0 +1,49 @@
+            (if (>= (get-bms-val 'bms-v-cell-max) bms-vmax-limit-start) (bms-set-fault 'BMS_CELL_OVER_VOLTAGE))
+            (if (<= (get-bms-val 'bms-v-cell-min) bms-vmin-limit-start) (bms-set-fault 'BMS_CELL_UNDER_VOLTAGE))
+            (if (>= (get-bms-val 'bms-temp-cell-max) bms-tmax-limit-start) (bms-set-fault 'BMS_CELL_OVER_TEMP))
+            (if (<= (get-bms-val 'bms-temp-cell-max) bms-tmin-limit-start) (bms-set-fault 'BMS_CELL_UNDER_TEMP))
```
Mistake here, you're reading `'bms-temp-cell-max` instead of `min`.

Relys replied:
No mistake. bms-temp-cell-min doesn't exist. It's not provided by can, hence why I originally removed the check. I think looking at highest cell is probably fine for under temp check of 0 deg. The pack isn't going to have that extreme of an differential between temperatures and if a cell goes under 0 deg before triggered it's not critical.

lukash replied:
Oh, shoot... Well, it even could in case of a split pack, so this is far from ideal. You think maybe contributing this to the FW would be a good idea, if we wanna do this check? Would make sense to me.

Until then, could you add a comment explaining this?

Relys replied:
Yeah I think that's a question for Vedder regarding how to expand CAN_PACKET_BMS_SOC_SOH_TEMP_STAT https://github.com/vedderb/bldc/blob/07128eaebbec7433a7dfb37c8a44fab0090425c7/bms.c#L116 I can leave a comment in the mean time.

Relys replied:
Added comment.

lukash commented on src/bms.lisp:
```diff
@@ -0,0 +1,49 @@
+            (if (>= (get-bms-val 'bms-temp-ic) bms-ic-tmax-limit-start) (bms-set-fault 'BMS_OVER_TEMP))
+        })
+        (apply ext-bms-set-fault bms-fault)
+        (yield 10000)
```
This was 50ms in last iteration, I asked to increase it to 100ms, now it's 10ms. Could we possibly go to 200ms? Are there any drawbacks to a longer delay?

Relys replied:
ah i mathed wrong. Nah 200ms is fine. I think it's reasonable as far as polling for bms feedback.

lukash commented on src/main.c:
```diff
@@ -671,7 +683,8 @@ static bool check_faults(data *d) {
     float input_voltage = VESC_IF->mc_get_input_voltage_filtered();
 
-    if (input_voltage < d->float_conf.tiltback_hv) {
+    if ((input_voltage < d->float_conf.tiltback_hv) &&
```
I know lisp people like brackets, but you don't need to add them here :smile:. `<` has higher precedence than `&&`. You can always check here: https://en.cppreference.com/w/c/language/operator_precedence

It's good to add brackets when it's not obvious, but this is one of the most common cases and everyone should know what's the order. The brackets add noise.

Relys replied:
ah i think dado added that check. i'll change the brackets

lukash commented on src/main.c:
```diff
@@ -2582,6 +2638,14 @@ static lbm_value ext_set_fw_version(lbm_value *args, lbm_uint argn) {
 
+// Called from Lisp to pass in the fault code of the bms.
+static lbm_value ext_bms_set_fault(lbm_value *args, lbm_uint argn) {
+    unused(argn);
```
I'm pretty sure you should check that `argn > 0`, otherwise you'll be accessing random memory.

Relys replied:
I can try, but I thought I had some sort of an issue with it not passing for some reason.

lukash replied:
That'd be pretty bad though :sweat_smile: this needs to work...

Relys replied:
Yeah I should be able to check with a spare BMS even though I don't have a working board atm. I'll get back on this. The problem with it accessing uninitialized memory would only be if the function is called without a parameter...which it's only called one time with the parameter passed, so honestly I don't see a scenario where this would happen.

Edit: Ok yes I resolved it. The problem was it was using apply instead of just calling the function in lisp directly. It works with the checks now.

lukash reviewed (changes requested):
Thanks for the update. I unresolved some comments which I felt weren't addressed well enough. I also think you missed 4 comments from the previous review because GitHub hid them... For those I made new comments in the current review, so you can just react to my current comments and ignore the old ones...

Sorry about the amount, but it is what it is :sweat_smile: 

And thanks for upstreaming the v-cell-min/max values, makes it cleaner on our side.

lukash commented:
And please add the `Feature: ` trailer to the commit description, that'll be turned into a changelog record (so phrase it that way) with an attribution to you.

Relys commented:
Added another update. I'm unable to test right now still, so lmk if there's any errors in lisp or something.

Relys commented:
Also just added support back in for 6.05 (since we don't have 6.06 API changes yet) after talking to @surfdado and Fungineers.

Relys commented:
Another change to allow disabling this feature by setting Motor CFG -> BMS vmin-limit-start, vmax-limit-start or tmax-limit-start to 0.0

lukash commented:
> Another change to allow disabling this feature by setting Motor CFG -> BMS vmin-limit-start, vmax-limit-start or tmax-limit-start to 0.0

Hmm. I'm not sure about this, what's the motivation? FWIW IMO it'd be better if you brought up significant changes like this with me before doing them. If I have an issue with them, it makes me feel bad to bring it up because I don't want to make you go back and forth.

You can turn this BMS alerting functionality off by setting BMS Type to None, right? So what's the motivation?

What'd make more sense to me is disabling that one particular check by setting the limit to 0 (but we don't have a config value for every check we do that'd allow us to disable them individually :disappointed:).

Relys commented:
> > Another change to allow disabling this feature by setting Motor CFG -> BMS vmin-limit-start, vmax-limit-start or tmax-limit-start to 0.0
> 
> Hmm. I'm not sure about this, what's the motivation? FWIW IMO it'd be better if you brought up significant changes like this with me before doing them. If I have an issue with them, it makes me feel bad to bring it up because I don't want to make you go back and forth.
> 
> You can turn this BMS alerting functionality off by setting BMS Type to None, right? So what's the motivation?
> 
> What'd make more sense to me is disabling that one particular check by setting the limit to 0 (but we don't have a config value for every check we do that'd allow us to disable them individually 😞).

"You can turn this BMS alerting functionality off by setting BMS Type to None, right? So what's the motivation?"

You assume too much haha. No that's not exposed. @surfdado mentioned wanting some way to disable this feature, and so this was the only solution I could come up with the variables exposed in Lisp. I could take a look and see if BMS Type is exposed on the C side maybe? Edit: I checked and nope.

Relys committed:
Feature: Add Support for BMS Tiltback

lukash commented on package.lisp:
```diff
@@ -8,6 +8,17 @@
 
+(def version_major (first (sysinfo 'fw-ver)))
+(def version_minor (second (sysinfo 'fw-ver)))
+(if (or (eq (first (trap (get-bms-val 'bms-v-cell-min))) 'exit-ok) (or (>= version_major 7) (and (>= version_major 6) (>= version_minor 5)))) {
```
This part of the condition: `(>= version_major 6)` should be just `(= version_major 6)`, I think I suggested this in the first place but thought I corrected myself later.

Also, remind me, why do we need both `(eq (first (trap (get-bms-val 'bms-v-cell-min))) 'exit-ok)` and (well, logically it's `or`) `(or (>= version_major 7) (and (>= version_major 6) (>= version_minor 5)))`?

lukash commented on src/bms.lisp:
```diff
@@ -0,0 +1,93 @@
+    (var vmin-limit)
+    (var vmax-limit)
+    (var tmax-limit)
+    (var tmin-limit -10)
```
Why did you silently change this to -10? When you change the semantics, could you **please** comment with an explanation?

lukash commented on src/bms.lisp:
```diff
@@ -0,0 +1,93 @@
+    (var config-update-time)
+    (var fault 0u32)
+    (var v-cell-support (eq (first (trap (get-bms-val 'bms-v-cell-min))) 'exit-ok))
+    (var bms-data-version-support (and (eq (first (trap (get-bms-val 'bms-data-version))) 'exit-ok) (> (get-bms-val 'bms-data-version) 0)))
```
What does `bms-data-version-support` represent? Seems improperly named to me... Might need an explanation comment.

lukash commented on src/bms.lisp:
```diff
@@ -0,0 +1,93 @@
+        (if (> (secs-since config-update-time) 1.0) {
+            (update-config)
+        })
+        (if (or (= vmin-limit 0.0) (= vmax-limit 0.0) (= tmax-limit 0.0)) {
```
What is this condition for? It doesn't look right to skip all the checks if one of them is 0.

lukash replied:
Ok, I obviously didn't even read the last few messages when I came back with this review. But my message kind-of confirms this is confusing. The code is already running here, this check is not like an ordinary enable switch (in addition to it being hooked to three or-ed values). IMO a global in-package switch (as I described in the review) would be much better. Potentially any of the above being zero could be used to disable that particular check (for when e.g. the user has improperly configured thermistors or some other malfunction of one them).

lukash commented on src/bms.lisp:
```diff
@@ -0,0 +1,93 @@
+                (if bms-data-version-support {
+                    (setq t-cell-min (get-bms-val 'bms-temps-adc 1))
+                    (setq t-cell-max (get-bms-val 'bms-temps-adc 2))
+                    (var t-mosfet-temp (get-bms-val 'bms-temps-adc 3))
```
What do the `1`, `2` and `3` represent here? Why do you set `t-cell-min` to `1` and `t-cell-max` to `2`?

lukash commented on src/bms.lisp:
```diff
@@ -0,0 +1,93 @@
+                    (setq t-cell-min (get-bms-val 'bms-temps-adc 1))
+                    (setq t-cell-max (get-bms-val 'bms-temps-adc 2))
+                    (var t-mosfet-temp (get-bms-val 'bms-temps-adc 3))
+                    (if (> t-mosfet-temp -280) {
```
What is the `-280` constant?

lukash commented on src/bms.lisp:
```diff
@@ -0,0 +1,93 @@
+                    })
+                } {
+                    (setq t-cell-min (get-bms-val 'bms-temp-cell-max))
+                    (setq t-cell-max t-cell-min)
```
I think this logic around the `t-cell-min` and `t-cell-max` needs some comments. Not clear what's going on.

lukash commented on src/main.c:
```diff
@@ -1443,6 +1488,17 @@ static void refloat_thd(void *arg) {
+            if (bms_is_fault_set(d->bms_fault, BMSF_CELL_BALANCE)) {
+                beep_alert(d, 3, true);
+                d->beep_reason = BEEP_CELL_BALANCE;
+            }
```
IIUC this is gonna be beeping pretty much continuously (not sure the 3 beeps actually make any difference, I'd say not). Did you test this?

lukash reviewed (commented):
Hey, I'm very sorry about the long delay. Read through it another time, this time I have more questions than anything...

In addition, I think we need a master switch for this functionality in the Refloat Cfg -> Specs tab. If disabled, just don't even initialize the lisp code. The feature as a whole also needs all the thresholds documented, this new option's help text could be a good place. That is, all hardcoded thresholds should be listed there and the options that are used from VESC BMS config should be listed there along with the details (e.g. having a 3 degree offset etc...).

And if the option is enabled, it should _only_ report `BMSF_NONE` if it's running properly and everything is ok. In case of unsupported FW version, it should report a special error, as well as in case of any other state than "all checks are ok". That way, when the user enabled the feature, they can be sure what's up with it. The way it is now, it seems to me you're kind-of hoping...

lukash commented:
Superseeded by #37.
